Update macOS frameworks to 10.12

matthewbauer opened the pull request:
###### Motivation for this change

Lots of updates to Darwin stuff are now possible following @copumpkin's update in #46704.

A few things I want to get to:

- Bump libSystem/XNU to 10.12
- Bump frameworks to 10.12 (or maybe 10.13 for consistency)
- ~Update Security/SecurityTool~

GrahamcOfBorg commented:
<!--REQUEST_ID=8782f15f-a31b-4986-bbcb-079f08eed880-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=e6e03dc4-60b6-481c-89b5-5d297a7a27b4)

Attempted: coreutils, krb5, patch

The following builds were skipped because they don't evaluate on aarch64-linux: darwin.hfs, darwin.xnu

<details><summary>Partial log (click to expand)</summary><p>

```
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


these paths will be fetched (0.84 MiB download, 3.07 MiB unpacked):
  /nix/store/nl80wx84gd0pl427niwgf47g580ywggv-krb5-1.15.2
copying path '/nix/store/nl80wx84gd0pl427niwgf47g580ywggv-krb5-1.15.2' from 'https://cache.nixos.org'...
/nix/store/8z916jw6wslyy089hr0s8mr0x75gjaxg-coreutils-8.30
/nix/store/nl80wx84gd0pl427niwgf47g580ywggv-krb5-1.15.2
/nix/store/a8ss5b93lpayri92gpzcnzxdbgqikdkn-patch-2.7.6
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=8782f15f-a31b-4986-bbcb-079f08eed880-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=49754198-6284-45ee-ad60-2ce38f160b30)

Attempted: coreutils, krb5, patch

The following builds were skipped because they don't evaluate on x86_64-linux: darwin.hfs, darwin.xnu

<details><summary>Partial log (click to expand)</summary><p>

```
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


these paths will be fetched (0.95 MiB download, 3.24 MiB unpacked):
  /nix/store/y2baad5sx4s250b6i94smivgiak7k5p2-krb5-1.15.2
copying path '/nix/store/y2baad5sx4s250b6i94smivgiak7k5p2-krb5-1.15.2' from 'https://cache.nixos.org'...
/nix/store/hr4y6cdbg56hc8b6vjdim5p7247gkk8z-coreutils-8.30
/nix/store/y2baad5sx4s250b6i94smivgiak7k5p2-krb5-1.15.2
/nix/store/8yhhh8h46yz20g6kqsxymzav48942115-patch-2.7.6
```
</p></details>

LnL7 commented on pkgs/os-specific/darwin/apple-source-releases/Libsystem/system_kernel_symbols:
```diff
@@ -790,6 +790,8 @@ _removexattr
 _rename_ext
 _renameat
+_renamex_np
+_renameatx_np
```
Are these available on 10.12 and earlier?

matthewbauer replied:
It looks like they were introduced in xnu-3789.70.16 which corresponds to 10.12. gnulib uses it since this commit:

https://github.com/coreutils/gnulib/commit/9abc08406d458ce18eb98a286aaae618a33dd231

and from updating to latest xnu:

https://github.com/apple/darwin-xnu/blob/0a798f6738bc1db01281fc08ae024145e84df927/bsd/sys/stdio.h#L47-L48

Here is the file that defines it in XNU:

http://newosxbook.com/src.jl?tree=xnu-3789.70.16&file=/libsyscall/wrappers/renamex.c

I wonder what people think about breaking support for 10.10 and 10.11? How common are they in the wild?

LnL7 replied:
Well 10.12 is pretty important because I'm pretty sure most builders are still running that. We probably don't have to keep older versions, but we should officially drop support for those then.

veprbl replied:
I was working on automating the generation of those symbol lists as a step towards #23432
https://github.com/veprbl/libsystem_symbol_list/branches <- tool to collect lists
https://github.com/veprbl/nixpkgs/commits/libsystem_symbols <- branch intended for the PR, includes the lists and scripts to take intersection
I was never able to dedicate CPU time to bootstrap this before making a PR, but if someone would be interested in picking this up, feel free :)

matthewbauer replied:
Yeah that definitely seems useful. We don't want to use the symbols from the latest libsystem just because it would break for older versions. I wonder if anyone is still running 10.12 who could run the script? Most likely there's only a few new symbols.

matthewbauer replied:
@veprbl Were there any issues with this method? Might be worth putting in this PR.

veprbl replied:
There is a 10.12.6 branch in libsystem_symbol_list that was acquired from Travis machine. It would be nice to have a volunteer to contribute a list for 10.14.

veprbl replied:
@matthewbauer Not that I know of. In the end of the day you just get a list and the rest of the machinery stays the same. You are not able to easily edit it, but it least its content have a clear motivation. And you also recover some missing symbols that used to be skipped in manually maintained list.

matthewbauer committed:
darwin: 10.11 → 10.12

Lots of stuff has gotten moved around. Many security libraries have been merged
into the Security monorepo. I’ve cleared them out for now, we will
need to modify Security to build them!

This also moves some things around to more clearly separate
bootstrapping the stdenv from everything else. We want the “normal”
mode to be the non-bootstrapped version. When you ask for “Security”,
you want the actual built software, not a crippled one.

- Add TARGET_OS_OSX to darwin.libSystem. Looks like something
  introduced in 10.12. TARGET_OS_MAC is only set when building for
  desktop (iOS will have TARGET_OS_MAC set)
- Bump darwin.dtrace
- Bump darwin.libpthread
- Remove SmartCardServices, libsecurity*, etc.
- Install some more headers for darling.

periklis commented:
This will enable finally modern tools like chunkwm to be packaged 💯

LnL7 commented:
Yeah, I think so.

matthewbauer committed:
xcbuild: set -Wno-error

gcc doesn’t know all of the specific flags here. It is easier to just
set -Wno-error.

copumpkin commented:
This is mostly great, but it doesn't feel like nixpkgs is really the place to keep the archive of all libsystem symbol versions, especially when those files aren't actually used in the evaluation of nixpkgs (just as input to a script that produces the final intersected list). Perhaps we could just shove them into a separate repo in the NixOS org?

Infinisil commented:
Also, this should probably go to staging

copumpkin commented:
Yeah, I'd probably send it to staging and maybe even trigger a PR job in Hydra for it, since it can have pretty far-reaching implications and it'd be nice to not be surprised by them.

LnL7 commented:
This shouldn't go to staging before we run a wip jobset, I want to have an idea of the impact before polluting staging with potentially problematic changes.

LnL7 commented:
What's the status here? The only regression I noticed when running a small set of builds was `darwin.libutil`.

matthewbauer commented:
I would be happy to grt this merged. I haven’t had time to really clean this up and stuff. Mostly someone will need to be around if there are any regressions. I think the only notable change this brings is that macOS 10.11 will no longer work properly.

I am still working on getting securitytool to build. But it’s gotten a lot harder than i expected.

LnL7 commented:
We should definitively move the symbols to a separate repo, but we don't have to wait for that to test it further. I would propose to rebase against a hydra eval (eg. 206a1c00baea8678dcce8c75dcc3df48ba59539b for [trunk 1490895](https://hydra.nixos.org/eval/1490895)) and start off a darwin eval.

LnL7 commented:
Oh, I missed the evaluation error `Libsystem/default.nix:1:1 called with unexpected argument 'hfs'`

Ericson2314 commented:
I like how the number of custom phases was cut down. Great work!

matthewbauer commented:
We may want to wait to merge this until someone can bump bootstrap-tools... I'm not sure if this will effect what macOS versions we can use.

GrahamcOfBorg commented:
<!--REQUEST_ID=416a0b5d-7f7f-47ca-a0b9-717194cc732b-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=4738aecf-499e-45db-b200-9dd8f7aee477)

Attempted: xcbuild

The following builds were skipped because they don't evaluate on aarch64-linux: darwin.dtrace, darwin.libpthread

<details><summary>Partial log (click to expand)</summary><p>

```
Call Stack (most recent call first):
  CMakeLists.txt:10 (project)


-- Configuring incomplete, errors occurred!
See also "/build/source/build/CMakeFiles/CMakeOutput.log".
See also "/build/source/build/CMakeFiles/CMakeError.log".
builder for '/nix/store/81dgz0fb5fl5v5ch9vp6vpv2qi2a7lrm-xcbuild-0.1.2-pre.drv' failed with exit code 1
cannot build derivation '/nix/store/gvg697aads0zzaw1plxq6w9bsn108yvg-xcodebuild-0.1.2-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/gvg697aads0zzaw1plxq6w9bsn108yvg-xcodebuild-0.1.2-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e9caeb7e-473b-4ad8-934d-45f0a55821c4-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=54b5a4de-b3fe-4793-a393-3f0f3fb47b15)

Attempted: xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre/bin/builtin-copy
shrinking /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre/bin/actool
strip is /nix/store/0z21vv6i3bg0p5k7pf0cg5qzv2psnpr7-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre/lib  /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre/bin
patching script interpreter paths in /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre
checking for references to /build/ in /nix/store/cm517nbmzvlx5rbh6iq43inacdjrcl49-xcbuild-0.1.2-pre...
building '/nix/store/qphwdw6q4i3i6mfg0yw4dad5f9dxyb56-xcodebuild-0.1.2-pre.drv'...
shrinking RPATHs of ELF executables and libraries in /nix/store/q92drb11mkrzb5w5y53z29cy01syxf9i-xcodebuild-0.1.2-pre
patching script interpreter paths in /nix/store/q92drb11mkrzb5w5y53z29cy01syxf9i-xcodebuild-0.1.2-pre
checking for references to /build/ in /nix/store/q92drb11mkrzb5w5y53z29cy01syxf9i-xcodebuild-0.1.2-pre...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e9caeb7e-473b-4ad8-934d-45f0a55821c4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=1a2e89da-f342-461b-b0fc-0394f56a0242)

Attempted: xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/224dlfp1pw63c4n530i4dcl7pss6r6s7-xcbuild-0.1.2-pre/lib  /nix/store/224dlfp1pw63c4n530i4dcl7pss6r6s7-xcbuild-0.1.2-pre/bin
patching script interpreter paths in /nix/store/224dlfp1pw63c4n530i4dcl7pss6r6s7-xcbuild-0.1.2-pre
checking for references to /build/ in /nix/store/224dlfp1pw63c4n530i4dcl7pss6r6s7-xcbuild-0.1.2-pre...
building '/nix/store/351skal78ljhp9xq5prns6vv0pm5j1bf-Platforms.drv'...
building '/nix/store/krnx7xgyvip3idf7ci694wgcqfpnkqrl-xcode-select.drv'...
building '/nix/store/864jqf16z3nc5y1b7pnbh373mir81ggp-xcrun.drv'...
building '/nix/store/l5fnynjkmzxqy17rzqqdbc8z8x4yhq35-xcodebuild-0.1.2-pre.drv'...
shrinking RPATHs of ELF executables and libraries in /nix/store/174x2873m4aam08w2q69v3dc3n2vacn6-xcodebuild-0.1.2-pre
patching script interpreter paths in /nix/store/174x2873m4aam08w2q69v3dc3n2vacn6-xcodebuild-0.1.2-pre
checking for references to /build/ in /nix/store/174x2873m4aam08w2q69v3dc3n2vacn6-xcodebuild-0.1.2-pre...
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=e9caeb7e-473b-4ad8-934d-45f0a55821c4-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=fe514a33-6e1d-45bc-82db-b4fb2b5d6a77)

Attempted: xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/8v4hbx47269sw9i5fda035xjwcjal5r6-apple-framework-Security.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/kx60p2f7dnpgsrbyz28nl71mpy0fp133-apple-framework-Accelerate.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/9sx9w4nq2k5pz7pibw36i5lpc3zb6fbk-apple-framework-ServiceManagement.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/jfd60wy579m00xk4vg3pq8q4zkgmlwkr-apple-framework-SystemConfiguration.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/i1j139l79wv2jg0s9aq108l4j5kij4dy-apple-framework-CoreGraphics.drv': 8 dependencies couldn't be built
cannot build derivation '/nix/store/mr7yjc81fsnk5vpcj892105aankzcd7c-apple-framework-CoreServices.drv': 12 dependencies couldn't be built
cannot build derivation '/nix/store/4bf2bxbangmms49hjncyknrmbmd29i7g-apple-framework-ImageIO.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/c549wg4pmkxj42qzgf0i15cqspma4qqn-xcbuild-0.1.2-pre.drv': 10 dependencies couldn't be built
cannot build derivation '/nix/store/7wlv3pq4hdi5whi439zh30vyzpip3bk4-xcodebuild-0.1.2-pre.drv': 9 dependencies couldn't be built
error: build of '/nix/store/7wlv3pq4hdi5whi439zh30vyzpip3bk4-xcodebuild-0.1.2-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=416a0b5d-7f7f-47ca-a0b9-717194cc732b-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=83bd3790-4a16-410b-ad9b-a9cd3a654b2c)

Attempted: xcbuild

The following builds were skipped because they don't evaluate on x86_64-linux: darwin.dtrace, darwin.libpthread

<details><summary>Partial log (click to expand)</summary><p>

```
Call Stack (most recent call first):
  CMakeLists.txt:10 (project)


-- Configuring incomplete, errors occurred!
See also "/build/source/build/CMakeFiles/CMakeOutput.log".
See also "/build/source/build/CMakeFiles/CMakeError.log".
builder for '/nix/store/i7dr5xx4gn01s5vg7qcz62f04w0y2zjp-xcbuild-0.1.2-pre.drv' failed with exit code 1
cannot build derivation '/nix/store/kwbwqxsv85cca4vz47faiywmvdhsh115-xcodebuild-0.1.2-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/kwbwqxsv85cca4vz47faiywmvdhsh115-xcodebuild-0.1.2-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=416a0b5d-7f7f-47ca-a0b9-717194cc732b-->
Timed out, unknown build status on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.47678&attempt_id=cec2abc0-f557-4a4d-bb2a-933ce6559b1d)

Attempted: darwin.dtrace, darwin.libpthread, xcbuild

<details><summary>Partial log (click to expand)</summary><p>

```
cannot build derivation '/nix/store/9sx9w4nq2k5pz7pibw36i5lpc3zb6fbk-apple-framework-ServiceManagement.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/jfd60wy579m00xk4vg3pq8q4zkgmlwkr-apple-framework-SystemConfiguration.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/i1j139l79wv2jg0s9aq108l4j5kij4dy-apple-framework-CoreGraphics.drv': 7 dependencies couldn't be built
cannot build derivation '/nix/store/mr7yjc81fsnk5vpcj892105aankzcd7c-apple-framework-CoreServices.drv': 11 dependencies couldn't be built
cannot build derivation '/nix/store/4bf2bxbangmms49hjncyknrmbmd29i7g-apple-framework-ImageIO.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/6gkmxvl6dpac5fp6bf8gwlnvs6qhysch-xcbuild-0.1.2-pre.drv': 8 dependencies couldn't be built
cannot build derivation '/nix/store/7d2cdaz560zm3aynvkykbg7yn3v8yngz-xcodebuild-0.1.2-pre.drv': 8 dependencies couldn't be built
cannot build derivation '/nix/store/z5c2px4a28kwlfq772vkqjc7qqb9nirs-hook.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/lmdv6vsbqfcixlwh7v8mfldkwg6z6319-dtrace-osx-10.12.6.drv': 11 dependencies couldn't be built
error: build of '/nix/store/7d2cdaz560zm3aynvkykbg7yn3v8yngz-xcodebuild-0.1.2-pre.drv', '/nix/store/90qarrc4jprs36znmh5qzy1369n1f4h0-libpthread-osx-10.12.6.drv', '/nix/store/lmdv6vsbqfcixlwh7v8mfldkwg6z6319-dtrace-osx-10.12.6.drv' failed
```
</p></details>

matthewbauer commented:
@grahamcofborg eval

matthewbauer commented:
@lnl7 this should be ready to go! I haven't looked at it in a while, so will need someone to go through this again just to make sure it doesn't mess anything up. I was originally planning to include fixes for "securitytool" in here - but haven't gotten it to work. But otherwise this should be safe to go to staging I think.

LnL7 commented:
Both coreutils and patch are failing for me, but this looks promising https://github.com/Homebrew/homebrew-core/commit/440b9e7bc457af83687843fea84a9f80f9642283. Looks like we didn't run into that yet because we keep the sdk versions stable.

```
Undefined symbols for architecture x86_64:
  "_renameatx_np", referenced from:
      _renameat2 in libpatch.a(renameat2.o)
ld: symbol(s) not found for architecture x86_64
```

matthewbauer commented:
Yes- You need to be running 10.12+ for this to work

matthewbauer commented:
Does that happen on 10.12??? I have a 10.11 machine that it breaks on but I was hoping that would be the only one broken.

LnL7 commented:
Yeah, this is 10.12. Does this not happen on newer versions?

matthewbauer commented:
Yeah - I'm surprised it happens on 10.12. The symbol was added by Apple recently:

http://newosxbook.com/src.jl?tree=xnu-3789.70.16&file=/bsd/sys/stdio.h

But it should be there for 10.12. Maybe it needs it listed in the symbols list?

matthewbauer commented:
Sorry the original commit was missing the new symbols to libSystem. That error should be fixed in 716f9fb.

LnL7 commented:
Like I mentioned before, this should be tested on hydra before merging to staging. It causes 8600 breakages and the stdenv is still broken on 10.12 and thus will block the new staging-next until fixed or reverted.

https://hydra.nixos.org/eval/1496326#tabs-now-fail

```
Undefined symbols for architecture x86_64:
  "_fs_snapshot_root", referenced from:
     -reexported_symbols_list command line option
  "_mk_timer_arm_leeway", referenced from:
     -reexported_symbols_list command line option
ld: symbol(s) not found for architecture x86_64
```

matthewbauer commented:
Sorry about that!

It's very weird that this is not happening locally. Most likely we can remove those entries from system_kernel_symbols safely. The symbols should be available on 10.12.6 but perhaps not on earlier 10.12.

LnL7 commented:
The VM was installed with 10.12.2, looks like it's much better after updating which is a bit unfortunate but not as bad as I thought. Can you look into the binutils issue? I'll start another eval when that's fixed to see what remains.
